e2e/nomostest/nt.go

```diff
@@ -752,123 +749,61 @@ func (nt *NT) MustDeleteGatekeeperTestData(file, name string) {
-		if pod.Name != nt.otelCollectorPodName {
-			ctx, cancel := context.WithCancel(nt.Context)
-			port, err := nt.ForwardToFreePort(ctx, ocmetrics.MonitoringNamespace, pod.Name,
-				testmetrics.OtelCollectorMetricsPort)
-			if err != nil {
-				cancel()
-				return err
-			}
-			nt.T.Cleanup(cancel) // stop port-forward
-			nt.otelCollectorPort = port
-			nt.otelCollectorPodName = pod.Name
-		}
-		return nil
-	})
-	if err != nil {
-		nt.T.Fatal(err)
-	}
-	nt.T.Logf("took %v to wait for otel-collector port-forward", took)
-}
-
-// PortForwardGitServer forwards the git-server deployment to a port.
-func (nt *NT) PortForwardGitServer() {
-	nt.T.Helper()
-
-	pod, err := nt.GetDeploymentPod(testGitServer, testGitNamespace)
-	if err != nil {
-		nt.T.Fatal(err)
-	}
-	podName := pod.Name
-
-	if podName != nt.gitRepoPodName {
+// portForwardOtelCollector forwards the otel-collector pod.
+func (nt *NT) portForwardOtelCollector() {
+	if nt.otelCollectorPortForward == nil {
 		ctx, cancel := context.WithCancel(nt.Context)
-		port, err := nt.ForwardToFreePort(ctx, testGitNamespace, podName, 22)
-		if err != nil {
-			cancel()
-			nt.T.Fatal(err)
-		}
-		nt.T.Cleanup(cancel) // stop port-forward
-		nt.gitRepoPort = port
-		nt.gitRepoPodName = podName
+		nt.T.Cleanup(cancel)
+		nt.otelCollectorPortForward = newPortForward(ctx, nt,
+			ocmetrics.MonitoringNamespace, ocmetrics.OtelCollectorName, fmt.Sprintf(":%d", testmetrics.OtelCollectorMetricsPort))
 	}
 }
 
-// ForwardToFreePort forwards a local port to a port on the pod and returns the
-// local port chosen by kubectl.
-func (nt *NT) ForwardToFreePort(ctx context.Context, ns, pod string, port int) (int, error) {
+// portForwardGitServer forwards the git-server deployment to a port.
+func (nt *NT) portForwardGitServer() {
 	nt.T.Helper()
-
-	nt.T.Logf("starting port-forward to pod %s/%s", ns, pod)
-	cmd := exec.CommandContext(ctx, "kubectl", "--kubeconfig", nt.kubeconfigPath, "port-forward",
-		"-n", ns, pod, fmt.Sprintf(":%d", port))
-
-	stdout := &strings.Builder{}
-	stderr := &strings.Builder{}
-	cmd.Stdout = stdout
-	cmd.Stderr = stderr
-
-	err := cmd.Start()
-	if err != nil {
-		return 0, err
-	}
-	if stderr.Len() != 0 {
-		return 0, fmt.Errorf(stderr.String())
-	}
-
-	// Log when the command exits
-	go func() {
-		if err := cmd.Wait(); err != nil {
-			nt.T.Logf("port-forward exited to pod %s/%s (%v)", ns, pod, err)
-		}
-	}()
-
-	localPort := 0
-	// In CI, 1% of the time this takes longer than 20 seconds, so 30 seconds seems
-	// like a reasonable amount of time to wait.
-	took, err := Retry(30*time.Second, func() error {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		default:
-			s := stdout.String()
-			if !strings.Contains(s, "\n") {
-				return fmt.Errorf(
-					"nothing written to stdout for kubectl port-forward to pod %s/%s, stdout=%s",
-					ns, pod, s)
+	if nt.gitRepoPortForward == nil {
```

Comment:
Don't like all these unlocked nil checks on the NT.
Can you just construct the PortForwarder/DeploymentProxy in test setup and then the places that need to use it can call nt.GitRepoProxy.LocalAddress() or nt.GitRepoProxy.LocalPort()?

Reply:
This is only called in the test setup. The nil check is just to be defensive in case someone added code which calls it a second time but I can remove it if you prefer

Reply:
If that's the case, make it return if not-nil (fast fail) and log a warning with why that's not necessary. Or maybe even Fatal, if we don't want it to happen.

Reply:
done